### ISSUE TITLE
Getting insufficient memory issue during csla serialization on large data.

### ISSUE THREAD

**Issue by cslalearner** [MarimerLLC/cslaforum#809]
**Question**
Getting insufficient memory issue during csla serialization on large data.

**Version and Platform**
CSLA version: v4.0.30319
OS: Windows
Platform: Asp.net,Silverlight

**Comment by rockfordlhotka** [comment]
You might compare the `DefaultBinding` from [WcfProxy in CSLA 4.5](https://github.com/MarimerLLC/csla/blob/9037cfee81f7fe98b1d8dec2150bc9962571f7e2/Source/Csla.Shared/DataPortalClient/WcfProxy.cs) to see if it is different from your current 4.0 version.

If it is different you can create a subclass of `WcfProxy` to supply your own default binding for WCF.

I don't remember exactly when we hard-coded the use of max values into the default binding, and if your version pre-dates that point then the WCF size limits might be the problem.

**Comment by cslalearner** [comment]
I have checked default-binding and it is implemented as in CSLA 4.5. Also wcf services are set to max limit "2147483647".
I am getting this error as soon as BeginSave hit. Is there any way I can debug that method?
I have also downloaded the source code and tried to change BusinessBase BeginSave method to add logging and add the reference of that csla in my project but when I run my project it points to old csla only. Could you please help me with this.

**Comment by rockfordlhotka** [comment]
Yes, you can debug the method, but it takes a little effort.

1. Remove all NuGet references to Csla packages (not just core, but all of them)
1. Add _existing project_ for the ".Shared" project(s) you are using (e.g. `Csla.Shared` and `Csla.Xaml.Shared`)
1. Add an existing project for `Csla.NetStandard2.0.csproj` - assuming that's your target
1. Add existing projects for other Csla packages (such as `Csla.Xaml.Xamarin`)

Your solution should now include the necessary CSLA projects. So _now_ you can go through _your_ projects and add _project references_ to the Csla projects used by each project.

At this point when you build/run/debug your code you can also walk through all the CSLA code.

**Comment by cslalearner** [comment]
Hello,
Thanks, I am able to debug the code and I have found that wcf proxy is not as implemented as in CSLA 4.5. My bad for wrong update, Wcf proxy is older thant csla 4.5. I wil take  the proxy code from 4.5 but could you please suggest me how I can update wcf service reference so that I can get the update wcf service.

**Comment by rockfordlhotka** [comment]
That server-side project was never part of the CSLA source code. Any time I updated the WCF service contract (which was not often) I did the following:

1. Build CSLA
1. Create a new ASP.NET project (in a totally new/different VS solution)
   1. Reference CSLA
   1. Create a dataportal.svc file 
   1. Set up necessary data portal configuration in `web.config`
   1. Run the project so the endpoint is alive in IIS Express
1. Open the Csla project in Visual Studio
   1. Remove the old service reference from the Csla project
   1. Remove the old service reference from the Csla.Silverlight project
   1. Add a new service reference in the Csla project against that temporary web site
   1. Add a new service reference in the Csla.Silverlight project against that temporary web site
   1. Save the solution
1. Stop, close, and delete the temporary web project

I probably would have automated that better, but the WCF endpoint contract changed so rarely it never seemed worth the effort.

**Comment by rockfordlhotka** [comment]
But do you really need to recreate the whole proxy/host pair? Didn't the `WcfProxy` type in 4.3 allow subclassing to override the method where the actual WCF proxy is created?

**Comment by rockfordlhotka** [comment]
Hmm, are you _sure_ you are using 4.3.14?

Here's the default proxy initialization from the [4.3.14 release](https://github.com/MarimerLLC/csla/releases/tag/V4.3.14)

```c#
    public static System.ServiceModel.Channels.Binding DefaultBinding
    {
      get 
      {
        if (_defaultBinding == null)
        {
          _defaultBinding  = new BasicHttpBinding();
          BasicHttpBinding binding = (BasicHttpBinding)_defaultBinding;
          binding.MaxBufferSize = int.MaxValue;
          binding.MaxReceivedMessageSize = int.MaxValue;
          binding.ReceiveTimeout = TimeSpan.FromMinutes(TimeoutInMinutes);
          binding.SendTimeout = TimeSpan.FromMinutes(TimeoutInMinutes);
          binding.OpenTimeout = TimeSpan.FromMinutes(TimeoutInMinutes);
        }; 
        return _defaultBinding; 
      }
      set { _defaultBinding = value; }
    }
```

In any case, it is true that in 4.3.13 you can override the `GetProxy` method to initialize your own WCF proxy instance:

```c#
    protected virtual WcfPortal.WcfPortalClient GetProxy()
```

Finally, I was just out walking my dog and it occurred to me that the max limit you are hitting **might be on the server** not the client. You said you encounter this in a save operation right? So it is possible that the client serializes the data, WCF starts transferring it to the server, and the server hits a limit and fails, causing the exception.

In other words, check your IIS, ASP.NET, and WCF configuration values on the server as per the [Data Portal Configuration book](https://store.lhotka.net/using-csla-4-data-portal-configuration) to make sure all those values are maxed out as well.

On the server it is more complex than the client, because back then all traffic flowed through IIS to ASP.NET and from ASP.NET to WCF. Thankfully these days things are easier, but sadly that does you no good 😏

**Comment by cslalearner** [comment]
Yeah I am using 4.3.1.4 release.
I have checked on my wcf services and on IIS so max content length is "2147483647" . and issue is coming whenever data is greater than 50 mb. is there any restriction  on csla wcf?Could you please let me know? Also what about this 
protected virtual WcfPortal.WcfPortalClient GetProxy()
What Am I supposed to change here?
Shall I call my wcf here instead if iwcfportal?

**Comment by rockfordlhotka** [comment]
CSLA just uses WCF. There is no such thing as "csla wcf".

We know that CSLA is not the problem because you were able to call `Clone` on your object graph and that did not fail. So CSLA (the `MobileFormatter`) is able to serialize and deserialize your object graph.

Also the exception detail you provided indicates the exception is coming from WCF, not from CSLA.

It might be the case that there is a limitation on the size of data Silverlight WCF can receive. I don't know the answer to that.

When you say you are using version 4.3.1.4, that might not be the same as 4.3.14. Especially because you said that, when you debugged the code, you don't think the WCF proxy object is being created using the settings from the version 4.3.14 'DefaultBinding' setup.

This is where you can create a subclass of `Csla.DataPortalClient.WcfProxy` and change the way the WCF proxy object is created by overriding the `GetProxy` method. You then configure your client to use your new proxy type instead of the one from CSLA.

This is described on page 29 of the _Data Portal Configuration_ book.

**Comment by GillesBer** [comment]
Hi,
If you want to debug only, without updating csla code, you can as an alternative:
- update the csla projects to generate 'full pdb'
- generate the NuGet package in prerelease mode
- reference the NuGet folder as a new NuGet feed/provider in your Visual studio
- upgrade your packages from this field

This is what we do in my team.
Regards

**Comment by rockfordlhotka** [comment]
A whole other approach might be to reduce the size of data on the wire. You can do that using the diffgram technique. This isn't something I've ever written up step-by-step, but there is a [Csla.DiffGram sample](https://github.com/MarimerLLC/csla/tree/master/Samples/Csla.DiffGram) showing how it can be done.

**Comment by cslalearner** [comment]
Ok,In WcfProxy I have set max value but it seems like error coming from Wcf Host. **Error says **"The remote server returned an error: NotFound."****
This type error, a wcf service can return if size of data exceeds from the allowed size .

Error coming in below method at return statement.
private System.IAsyncResult OnBeginUpdate(object[] inValues, System.AsyncCallback callback, object asyncState) {
            Csla.WcfPortal.UpdateRequest request = ((Csla.WcfPortal.UpdateRequest)(inValues[0]));
            return ((Csla.WcfPortal.IWcfPortal)(this)).BeginUpdate(request, callback, asyncState);
 }
I am sorry as I don't have much experience in CSLA my queries can be silly.
Could you please provide an insight on this?

**Comment by rockfordlhotka** [comment]
Not silly at all. WCF is really hard to configure correctly. The issues you are facing having nothing to do with CSLA - only WCF.

On the server there are several size limits that come from

1. IIS
1. ASP.NET
1. WCF

You need to make sure all three are set to allow maximum sizes for requests from the client. Also, WCF has more than one size limit you need to set.

Have you looked at the _Data Portal Configuration_ book? It has examples of how to set all these values.

**Comment by cslalearner** [comment]
Yeah, I have checked all the configuration mentioned  in data portal book & all are set for maximum value.So I thought to implement decompression as mentioned in your book. I have created a class compression and copy below method from the code mentioned in book:

protected override Csla.WcfPortal.CriteriaRequest ConvertRequest( Csla.WcfPortal.CriteriaRequest request) { Csla.WcfPortal.CriteriaRequest returnValue = new Csla.WcfPortal.CriteriaRequest(); returnValue.ClientContext = **CompressionUtility**.Compress(request.ClientContext); returnValue.GlobalContext = CompressionUtility.Compress(request.GlobalContext); if (request.CriteriaData != null) returnValue.CriteriaData = CompressionUtility.Compress(request.CriteriaData); returnValue.Principal = CompressionUtility.Compress(request.Principal); returnValue.TypeName = request.TypeName; return returnValue; }
But I have not found this compressionutility code or reference and I am not able to proceed further,could you please suggest from where I can get the code?

**Comment by cslalearner** [comment]
Actually ,I have implemented the compression and decompression and issue is fixed with it.But I want to know one more thing that can we do compression/decompression for specific request.like in my App.xaml.cs I have written like this:
  // use compressed proxy
 private void Application_Startup(object sender, StartupEventArgs e)
        {
            Csla.DataPortal.ProxyTypeName = typeof(Aztec.Backstage.App.Fusion.Pres.Compression.CompressionProxy<>).AssemblyQualifiedName;
}
So it is doing compression and decompression for all the request.How Can I do it for specific requests?

**Comment by rockfordlhotka** [comment]
@cslalearner if that's a goal then you need some code that'll look at the business type and decide whether to invoke the normal `HttpProxy` or your `CompressionProxy`.

That code is implemented as a custom proxy. Don't subclass `HttpProxy`, just directly implement `IDataPortalProxy` and add in your detection/switching code to route the call to the correct proxy object.

Your top-level configuration for the app will point to your new `RoutingProxy` (or whatever you call it).

**Comment by rockfordlhotka** [comment]
Hey @cslalearner , what compression libredid you end up using ?

**Comment by cslalearner** [comment]
Hello Rock,

I have used SharpZipLib for compression and decompression. Is that good or there is anything better I can use?

**Comment by rockfordlhotka** [comment]
I have no idea whether there's anything better, I was just curious as to what you ended up using. Thank you for sharing!
